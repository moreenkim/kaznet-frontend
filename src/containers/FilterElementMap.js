/// This Component takes an object of Filter Fields,
/// an array of isOpen States & an array of handleChange functions for
/// Filter Fields arranged in order according to the Filter Field Object values,
/// an optional array of strings and optional array of Filter Fields Position and returns the Filter Fields
/// as a Dropdown wrapped around the provided HTML tag optionally combined
/// with the array of strings ordered in specified Filter Fields Positon wrapped around its provided HTML tag
import React, { Component } from "react";
import { DropdownToggle, DropdownMenu, Dropdown } from "reactstrap";
import PropTypes from "prop-types";
import ElementMap from "./ElementMap";

export default class FilterElementMap extends Component {
  // Sets runtime typechecking for Component Props
  // Ensure props passed to FilterElementMap are of right type or Present
  static propTypes = {
    filterFields: PropTypes.object.isRequired,
    isOpenStates: PropTypes.arrayOf(PropTypes.bool).isRequired,
    handleChangeFunctions: PropTypes.arrayOf(PropTypes.func).isRequired,
    HTMLTag: PropTypes.string.isRequired,
    filterItemPositions: PropTypes.arrayOf(PropTypes.number),
    listItems: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const filterFields = this.props.filterFields;
    const filterFieldsList = [];
    const itemList = [];
    const isOpenStates = this.props.isOpenStates;
    const handleChangeFunctions = this.props.handleChangeFunctions;

    Object.entries(filterFields).map(([name, dropDownItems], index) => {
      filterFieldsList.push(
        <Dropdown
          key={index}
          isOpen={isOpenStates[index]}
          toggle={handleChangeFunctions[index]}
        >
          <DropdownToggle caret tag="span" className="nav-link">
            {name}
          </DropdownToggle>
          <DropdownMenu>{dropDownItems}</DropdownMenu>
        </Dropdown>
      );

      return null;
    });

    if (this.props.listItems) {
      if (this.props.filterItemPositions) {
        const filterItemPositions = this.props.filterItemPositions;
        itemList.push(...this.props.listItems);

        for (var i = 0; i < filterItemPositions.length; i++) {
          itemList.splice(filterItemPositions[i], 0, filterFieldsList[i]);
        }
      } else {
        itemList.push(...filterFieldsList);
        itemList.push(...this.props.listItems);
      }
    } else {
      itemList.push(filterFieldsList);
    }

    return (
      <ElementMap
        items={itemList}
        HTMLTag={this.props.HTMLTag}
        className={this.props.className}
      />
    );
  }
}
