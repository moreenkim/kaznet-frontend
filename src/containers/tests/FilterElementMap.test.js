// Test FilterElementMap
import React from "react";
import { shallow, mount } from "enzyme";
import FilterElementMap from "../FilterElementMap";
import toJson from "enzyme-to-json";

describe("containers/ElementMap", () => {
  it("renders without crashing", () => {
    const filterFields = {
      filterFields: []
    };
    const isOpenStates = [false];
    const handleChangeFunctions = [function() {}];

    shallow(
      <FilterElementMap
        filterFields={filterFields}
        isOpenStates={isOpenStates}
        handleChangeFunctions={handleChangeFunctions}
        HTMLTag="div"
      />
    );
  });

  it("renders tags correctly", () => {
    const filterFields = {
      filterFields: []
    };
    const isOpenStates = [false];
    const handleChangeFunctions = [() => {}];

    const wrapper = mount(
      <FilterElementMap
        filterFields={filterFields}
        isOpenStates={isOpenStates}
        handleChangeFunctions={handleChangeFunctions}
        HTMLTag="div"
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
