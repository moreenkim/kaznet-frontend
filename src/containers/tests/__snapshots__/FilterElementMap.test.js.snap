// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/FilterElementMap renders tags correctly 1`] = `
<FilterElementMap
  HTMLTag="div"
  filterFields={
    Object {
      "filterFields": Array [],
    }
  }
  handleChangeFunctions={
    Array [
      [Function],
    ]
  }
  isOpenStates={
    Array [
      false,
    ]
  }
>
  <ElementMap
    HTMLTag="div"
    items={
      Array [
        Array [
          <Dropdown
            active={false}
            addonType={false}
            direction="down"
            inNavbar={false}
            isOpen={false}
            nav={false}
            setActiveFromChild={false}
            toggle={[Function]}
      >
            <DropdownToggle
                  aria-haspopup={true}
                  caret={true}
                  className="nav-link"
                  color="secondary"
                  tag="span"
            >
                  filterFields
            </DropdownToggle>
            <DropdownMenu
                  flip={true}
                  tag="div"
            >
                  
            </DropdownMenu>
      </Dropdown>,
        ],
      ]
    }
  >
    <div
      key="0"
    >
      <Dropdown
        active={false}
        addonType={false}
        direction="down"
        inNavbar={false}
        isOpen={false}
        key="0"
        nav={false}
        setActiveFromChild={false}
        toggle={[Function]}
      >
        <Manager
          className="dropdown"
          onKeyDown={[Function]}
          tag="div"
        >
          <div
            className="dropdown"
            onKeyDown={[Function]}
          >
            <DropdownToggle
              aria-haspopup={true}
              caret={true}
              className="nav-link"
              color="secondary"
              tag="span"
            >
              <Target
                aria-expanded={false}
                aria-haspopup={true}
                className="nav-link dropdown-toggle"
                component="span"
                onClick={[Function]}
              >
                <span
                  aria-expanded={false}
                  aria-haspopup={true}
                  className="nav-link dropdown-toggle"
                  onClick={[Function]}
                >
                  filterFields
                </span>
              </Target>
            </DropdownToggle>
            <DropdownMenu
              flip={true}
              tag="div"
            >
              <div
                aria-hidden={true}
                className="dropdown-menu"
                role="menu"
                tabIndex="-1"
              />
            </DropdownMenu>
          </div>
        </Manager>
      </Dropdown>
    </div>
  </ElementMap>
</FilterElementMap>
`;
